### PR TITLE
For #475: implemented sunburn card with new feed logic

### LINKED ISSUE #475: Create Astromancer card Sunburn, its effects and tests
**Sunburn** 
Cost: 2 Mana 
Effect: Deal 12 to all unshielded characters. 
Type: Offensive
Feed-Lunar: 20% Chance to recast.

### PR DESCRIPTION
**Pull request by paulodamaso**
Closes #475: implemented sunburn card with new feed logic

### CODE REVIEW

File: src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+    public ApplyOnShieldBelowValue(Effect effect, int value) : base(effect)
+    {
+        _value = value;
+    }
+
+    public override bool Condition()
+    {
+        Debug.Log("Value: " + _value);
+        return Target.Members.Any(e => e.State[TemporalStatType.Shield] < _value );
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs]: Is this being applied to only the Unshielded members? The target group is going to be All for this one.

Comment (paulodamaso) [reply src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs]: @SilasReinagel This effect is about appling only to unshielded members, why should we apply it to all targets?

File: src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    private int _value;
+
+    public ApplyOnShieldBelowValue(Effect effect, int value) : base(effect)
+    {
+        _value = value;
+    }
+
+    public override bool Condition()
+    {
+        Debug.Log("Value: " + _value);
```
[the remark this answers is not in the document]
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs]: @SilasReinagel Killed

**Comment by paulodamaso** [comment]
> The recasting effect in the card has a duplication of the First Effect inside the second. This will be a maintainability nightmare. There should be a way to reference an Effect...

@SilasReinagel I agree that in this case the recasting effect can cause a maintainability problem, but I don't see a simple way of making effects reference other effects in the same card due to how we designed the effects fields to work on unity editor.

Explaining better, when we design cards in unity editor we don't have use the effect as unity prefab objects.  Instead, we set effect data and the parameters, and they are created in runtinme by our card logic. This allows us to compose the effects in a more or less efficient way without creating the `Inspector` pane on unity editor for each effect , but brings us some drawbacks, like we have a finite number of card actions and our effect composition is limited (I can't have conditional effects that trigger one effect if the condition is satisfied and another if not, for example).

So, I think the only way to allow us to make effects the way you proposed would be to add each effect as a unity object, and create all the user interface needed so we could use the same instance of some effect twice in the card.

In my opinon the time we will spend  looking for a solution for this will be wasted since we don't have any other card with the "recast on chance" effect on these conditions.

File: src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs (hunk 1 above)
Comment (SilasReinagel) [reply src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs]: @paulodamaso The question wasn't answered.

Are you certain that this correctly applies to only Unshielded members, even with a Group Target (All/All as per the card), even if one shielded member is present?

The logic in this condition looks wrong.

**Comment by SilasReinagel** [comment]
@paulodamaso More importantly, the card Design sets up the Recast for the NEXT card, not the effect on this card. The card is not correctly implementing the Recast for the NEXT card (Feed-Lunar)

File: src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs (hunk 1 above)
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Features/Effects/ApplyOnShieldBelowValue.cs]: @SilasReinagel Actually not:
`PlayedCard.perform()` calls `Apply` for each `Action` it has, for each target. But it dismantle the `Target` array and calls apply for each individual target. That's why `Target.Members.Any`, in this case (inside an effect), will ever be an individual target.

**Comment by paulodamaso** [comment]
@SilasReinagel I thought that we should recast this card effect, not the next one.

**Comment by paulodamaso** [comment]
@SilasReinagel Remade the effect implementation. Now we execute the recast card effect reactively.

File: src/ProjectNeon/Assets/Features/Battle/Cards/CardResolutionZone.cs [hunk 3]
```diff
@@ -9,6 +9,7 @@ public class CardResolutionZone : ScriptableObject
     [SerializeField] private CardPlayZone physicalZone;
     [SerializeField] private CardPlayZone playedDiscardZone;
     [SerializeField] private GameEvent onFinished;
+    [SerializeField] private BattleState battleState;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Battle/Cards/CardResolutionZone.cs]: This is a circular reference. You can not depend on something which depends on you.

File: src/ProjectNeon/Assets/Features/Effects/Events/ReplayLastCardBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿using UnityEngine;
+
+public class ReplayLastCardBehavior : MonoBehaviour
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/Events/ReplayLastCardBehavior.cs]: This should be attached to an object in the Battle Engine in the actual game.

**Comment by paulodamaso** [comment]
@SilasReinagel ping

**Pull request merged**